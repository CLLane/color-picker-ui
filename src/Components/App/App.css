body {
   background-image: url('https://i.pinimg.com/originals/47/cc/03/47cc0364b00fda0ebf63b2ae641c0945.png');
}

main {
  text-align: center;
}
p {
  margin: 0px;
}
h1 {
  margin: 0;
}
.header-container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
}

.all-palettes__button {
  margin-bottom: 30px;
}
.log-out__button {
  margin-bottom: 30px;
}
.logo__image {
  margin-left: 245px;
  margin-right: 230px;
  margin-bottom: 30px;
}