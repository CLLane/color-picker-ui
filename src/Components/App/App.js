import React, { Component } from "react";
import LoginForm from "../LoginForm/LoginForm";
import { ColorContainer } from "../ColorContainer/ColorContainer";
import { generateHexCode } from "../../utilities/helpers";
import "./App.css";
import  ProjectForm  from "../ProjectForm/ProjectForm";
import  ProjectsContainer  from "../ProjectsContainer/ProjectsContainer";
import PaletteContainer from "../PaletteContainer/PaletteContainer";
import { Route, Redirect, Link } from 'react-router-dom';
import {
  getUser,
  getUserProjects,
  getUserPalettes,
  postNewUser,
  postNewProject,
  postNewPalette,
  deletePalette,
  deleteProject,
  getAllPalettes
} from "../../utilities/apiCalls";

export class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true,
      error: "",
      colors: [
        { color: generateHexCode(), locked: false },
        { color: generateHexCode(), locked: false },
        { color: generateHexCode(), locked: false },
        { color: generateHexCode(), locked: false },
        { color: generateHexCode(), locked: false }
      ],
      user: null,
      user_projects: [],
      user_palettes: [],
      currentProjec: null,
      palettes: []
    };
  }

  loginUser = async (email, password) => {
    const userInfo = { email, password };
    try {
      const userId = await getUser(userInfo);
      if (userId) {
        this.setState({
          user: {
            id: userId,
            name: email.split("@")[0]
          }
        });
        this.userProjects(userId);
      }
    } catch (error) {
      this.setState({ error: error.message });
    }
  };

  userProjects = async userId => {
    try {
      const projects = await getUserProjects(userId);
      this.setState({ user_projects: projects });
      this.userPalettes(projects);
    } catch (error) {
      this.setState({ error: error.message });
    }
  };

  userPalettes = async projects => {
    try {
      let palettePromises = projects.map(async project => {
        const { id } = project;
        const palette = await getUserPalettes(id);
        return palette;
      });
      const palettes = await Promise.all(palettePromises);
      this.setState({ user_palettes: palettes.flat() });
    } catch (error) {
      this.setState({ error: error.message });
    }
  };

  signUpUser = async (email, password) => {
    const userInfo = { email, password };
    try {
      const userId = await postNewUser(userInfo);
      if (userId) {
        this.setState({
          user: {
            id: userId,
            name: email.split("@")[0]
          }
        });
      }
    } catch (error) {
      this.setState({ error: error.message });
    }
  };

  generateColors = () => {
    let { colors } = this.state;
    colors = colors.map(color => {
      return color.locked ? color : { color: generateHexCode(), locked: false };
    });
    this.setState({ colors });
  };


  toggleColorLock = (index) => {
    let { colors } = this.state;
    colors[index].locked = !colors[index].locked;
    this.setState({ colors })
  };

  handleSubmission = async (projectName, paletteName) => {
    const { colors, user, currentProject, user_projects } = this.state;
    const hex_codes = colors.map(colorObj => colorObj.color).join();
    if(user_projects.map(project => project.name).includes(projectName)) {
      const newPalette = { project_id: currentProject.id, hex_codes, name: paletteName };
      await this.createPalette(newPalette);
      this.userProjects(user.id);
    } else {
      const project_id = await this.createProject({ user_id: user.id, name: projectName });
      const newPalette = { project_id, hex_codes, name: paletteName };
      await this.createPalette(newPalette);
      this.userProjects(user.id);
    }
  }

  createProject = async (projectInfo) => {
    const projectId = await postNewProject(projectInfo);
    return projectId;
  }

  createPalette = async (paletteInfo) => {
    const paletteId = await postNewPalette(paletteInfo);
    return paletteId;
  }

  logoutUser = () => {
    this.setState({user: null})
  }

  trashPalette = async (id) => {
    const { user } = this.state;
    try {
      await deletePalette(id);
      await this.userProjects(user.id);
    } catch (error) {
      this.setState({error: error.message})
    }
  }

  trashProject = async (id) => {
    const { user } = this.state;
    try {
      await deleteProject(id);
      await this.userProjects(user.id);
      this.setState({ currentProject: null });
    } catch (error) {
      this.setState({error: error.message})
    }
  }

  allPalettes = async () => {
    try {
      const palettes = await getAllPalettes();
      this.setState({ palettes })
    } catch (error) {
      this.setState({ error: error.message })
    }
  }

  clearError = () => {
    this.setState({ error: '' });
  }

  updateCurrentProject = (project) => {
    this.setState({currentProject: project})
  }

  render() {
    const { error, user, colors, user_projects, user_palettes, currentProject, palettes} = this.state;
    return (
      <main>
        <Route path='/login' render={() => user ? <Redirect to='/'/> : (
          <>
            <img
              src="https://fontmeme.com/permalink/191011/986cec0ee62e1c276228d73cd566bdd2.png"
              alt="graffiti-fonts"
              border="0"
            />
            <LoginForm
              error={error}
              clearError={this.clearError}
              loginUser={this.loginUser}
              signUpUser={this.signUpUser}
            />
          </>
          )}/>
        <Route exact path='/' render={() => !user ? <Redirect to='/login' /> : (
            <>
            <h1>Welcome, {user.name}</h1>
            <button onClick={this.logoutUser}>Log Out</button>
            <Link to='/palettes'><button onClick={this.allPalettes}>Browse All Palettes</button></Link>
            <img
              src="https://fontmeme.com/permalink/191011/5ed4a0d9bcac8d65b68b8a1346771b36.png"
              alt="graffiti-fonts"
              border="0"
            />
            <ColorContainer
              colors={colors}
              generateColors={this.generateColors}
              toggleColorLock={this.toggleColorLock}
            /> 
            <ProjectForm 
              updateCurrentProject={this.updateCurrentProject}
              currentProject={currentProject}
              handleSubmission={this.handleSubmission}
              projects={user_projects}
            />
            <ProjectsContainer 
              projects={user_projects}
              palettes={user_palettes}
              trashPalette={this.trashPalette}
              trashProject={this.trashProject}
            />
            </>
          )
        } />
         <Route exact path='/palettes' render={() => !user ? <Redirect to='/login' /> : (
            <>
            <h1>Welcome, {user.name}</h1>
            <button onClick={this.logoutUser}>Log Out</button>
            <Link to='/'><button >Generate New Palette</button></Link>
            <img
              src="https://fontmeme.com/permalink/191011/5ed4a0d9bcac8d65b68b8a1346771b36.png"
              alt="graffiti-fonts"
              border="0"
            />
            <PaletteContainer 
              palettes={palettes}
            />
            <ProjectForm 
              updateCurrentProject={this.updateCurrentProject}
              currentProject={currentProject}
              handleSubmission={this.handleSubmission}
              projects={user_projects}
            />
            <ProjectsContainer 
              projects={user_projects}
              palettes={user_palettes}
              trashPalette={this.trashPalette}
              trashProject={this.trashProject}
            />
            </>
          )
        } />
      </main>
    );
  }
}

export default App;
