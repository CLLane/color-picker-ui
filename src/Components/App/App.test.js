import React from 'react';
import { shallow, maount } from 'enzyme';
import { MemoryRouter } from 'react-router'
import App from './App';
import './App.css';

describe('App', () => {
  let wrapper;
  let mockUser;
  let mockProject;

  beforeEach(() => {
    wrapper = shallow(<App />);
    mockUser = {
      id: 1, 
      name: 'Bob'
    };
    mockProject = { id: 10, name: 'Cool project'};
    wrapper.setState({user: mockUser, currentProject: mockProject, user_projects: [mockProject] })
    window.fetch = jest.fn();
  });


  it('Should match the snapshot with the proper data passed in.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should start with an array of 5 random colors', () => {
    expect(wrapper.state('colors').length).toEqual(5);
  });

  it('should be able to login a user and set the user to state', async () => {
    const mockUserInfo = { email: 'bob@gmail.com', password: 12345}
    const options = {
      method: 'POST',
      body: JSON.stringify(mockUserInfo),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = 'http://localhost:3001/user/login';
    wrapper.instance().loginUser('bob@gmail.com', 12345);
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should fetch palettes with the proper project id', async () => {
    window.fetch = jest.fn();
    await wrapper.instance().userPalettes([{id: 1}]);
    expect(window.fetch).toHaveBeenCalled();
  });

  it('should fetch projects with the proper user id', async () => {
    await wrapper.instance().userProjects(1);
    const url = 'http://localhost:3001/projects/1';
    expect(window.fetch).toHaveBeenCalledWith(url);
  });

  it('should try to sign up a user', async () => {
    const mockUserInfo = { email: 'bob@gmail.com', password: 12345}
    const options = {
      method: 'POST',
      body: JSON.stringify(mockUserInfo),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = 'http://localhost:3001/user/signup';
    wrapper.instance().signUpUser('bob@gmail.com', 12345);
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should generate a new random set of colors', () => {
    const oldColors = wrapper.state('colors');
    wrapper.instance().generateColors();
    const newColors = wrapper.state('colors');
    expect(newColors).not.toEqual(oldColors);
  });

  it('should toggle the locked property of a color, given the index', () => {
    const colorObj = wrapper.state('colors')[0];
    const expected = {...colorObj, locked: true};
    wrapper.instance().toggleColorLock(0);
    const updatedColorObj = wrapper.state('colors')[0];
    expect(updatedColorObj).toEqual(expected);
  });

  it('should try to create a new project given info', () => {
    wrapper.instance().createProject({name: 'test'});
    const options = {
      method: 'POST',
      body: JSON.stringify({name: 'test'}),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = 'http://localhost:3001/projects';
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should try to create a new palette given info', () => {
    wrapper.instance().createPalette({name: 'test'});
    const options = {
      method: 'POST',
      body: JSON.stringify({name: 'test'}),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = 'http://localhost:3001/palettes';
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('it should logout a user', () => {
    wrapper.setState({user: {id: 1, name: 'Bob'}})
    wrapper.instance().logoutUser();
    expect(wrapper.state('user')).toEqual(null);
  });

  it('should try to delete a palette, given an id', async () => {
    const options = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = `http://localhost:3001/palettes/1`;
    await wrapper.instance().trashPalette(1);
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should try to delete a project, given an id', async () => {
    wrapper.setState({user: {id:2, name: 'Bob'}});
    const options = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = `http://localhost:3001/projects/1`
    await wrapper.instance().trashProject(1);
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should try to get all palettes', async () => {
    const url = 'http://localhost:3001/palettes'
    await wrapper.instance().allPalettes();
    expect(window.fetch).toHaveBeenCalledWith(url);
  })

  it('should update the current colors array, given a palette', () => {
    const mockPalette = {
      id: 1, 
      name: 'Cool Colors',
      project_id: '12345',
      colorOne: '#123456',
      colorTwo: '#123456',
      colorThree: '#123456',
      colorFour: '#123456',
      colorFive: '#123456'
    };
    const expected = [
      {color: '#123456', locked: true},
      {color: '#123456', locked: true},
      {color: '#123456', locked: true},
      {color: '#123456', locked: true},
      {color: '#123456', locked: true},
    ];
    wrapper.instance().grabPalette(mockPalette);
    expect(wrapper.state('colors')).toEqual(expected);
  });

  it('it should clear the error message', () => {
    wrapper.setState({error: 'test error'});
    wrapper.instance().clearError();
    expect(wrapper.state('error')).toEqual('');
  });

  it('should set the given project to state', () => {
    const mockProject = {id: 1, name: 'Cool project'};
    wrapper.instance().updateCurrentProject(mockProject);
    expect(wrapper.state('currentProject')).toEqual(mockProject);
  });

  it('should try to edit the name of a project', () => {
    const mockProject = { id: 2, name: 'New' };
    const options = {
      method: 'PATCH',
      body: JSON.stringify(mockProject),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = `http://localhost:3001/projects/2`;
    wrapper.instance().updateProjectName(mockProject);
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should try to edit the name of a palette', () => {
    const mockPalette = { id: 2, name: 'New' };
    const options = {
      method: 'PATCH',
      body: JSON.stringify(mockPalette),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const url = `http://localhost:3001/palettes/2`;
    wrapper.instance().updatePaletteName(mockPalette);
    expect(window.fetch).toHaveBeenCalledWith(url, options);
  });

  it('should check if a project exists on the user before adding a palette', async () => {
    const colors = wrapper.state('colors');
    const hex_codes = colors.map(colorObj => colorObj.color).join();
    const mockPaletteInfo = { project_id: 10, hex_codes, name: 'Cool Palette' };
    wrapper.instance().handleSubmission('Cool project', 'Cool Palette');
      const options = {
        method: 'POST',
        body: JSON.stringify(mockPaletteInfo),
        headers: {
          'Content-Type': 'application/json'
        }
      };
      const newPaletteUrl = 'http://localhost:3001/palettes';
      expect(window.fetch).toHaveBeenCalledWith(newPaletteUrl, options);
  });

  it('should create a new project before adding a palette, if the project does not exist', () => {
    const mockUser = wrapper.state('user');
    const mockProjectInfo = {
      user_id: mockUser.id,
      name: 'uncool project'
    };
    const options = {
      method: 'POST',
      body: JSON.stringify(mockProjectInfo),
      headers: {
        'Content-Type': 'application/json'
      }
    };
    const projectUrl = 'http://localhost:3001/projects';
    wrapper.instance().handleSubmission('uncool project', 'uncool palette');
    expect(window.fetch).toHaveBeenCalledWith(projectUrl, options);
  })

});
