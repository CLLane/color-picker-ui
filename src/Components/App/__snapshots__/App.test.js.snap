// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should match the snapshot with the proper data passed in. 1`] = `
<main>
  <Route
    path="/login"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/palettes"
    render={[Function]}
  />
</main>
`;

exports[`App should match the snapshot when a user is present 1`] = `
<main>
  <Route
    path="/login"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/palettes"
    render={[Function]}
  />
</main>
`;
